text/0000-system-receptionist.md

```diff
@@ -0,0 +1,255 @@
+
+
+## Detailed design
+
+### Technical Background
+
+In both the actor model and XState, any actor ref can communicate with another actor ref as long as it has its `ActorRef` reference via `sendTo(actorRef, ...)`. It is possible to already implement the receptionist pattern in XState, but this involves a lot of wiring "boilerplate" and finding a way to pass through the receptionist actor ref to all spawned/invoked actors. This can be unfeasible for many use-cases, as it involves a lot of restructuring to accommodate the ad-hoc receptionist actor ref.
+
+Instead, we seek a built-in way to define a _system_ which will define a root machine and hold shared data (such as a registry of actor refs) for all of its actors, as well as abstractions to interact with the `system.receptionist` for registering and discovering system-wide actor refs. 
+
+
+### Implementation
+
+
+**Creating a system**
+
+The `createSystem(rootMachine, options?)` function will create an `ActorRef` whose behavior is defined by the `rootMachine`, which is a normal machine created from `createMachine(...)`. The purpose of the system is to provide `{ system }` to all machines (including the root) spawned in this system, which contains system-wide resources.
+
+```ts
+import { createSystem, createMachine } from 'xstate';
+
+const rootMachine = createMachine({
+  // ...
+});
+
+// Creating a system
+const system = createSystem(rootMachine);
+
+// Starting a system
+system.start();
+```
+
+**Registering to the receptionist**
+
+The `register(key)` action is made available from `'xstate/actions'`, which will register the executing actor to the system-wide registry under that `key`. Multiple actors can be registered under the same `key` by default, and it should be assumed that each `key` in the registry has 0 or more actor refs.
+
+```ts
+import { register } from 'xstate/actions';
+
+const todoMachine = createMachine({
+  // Register actor to the 'todo' key
+  entry: register('todo'),
+  // ...
+});
+```
+
+This should be roughly equivalent to:
+```ts
+// PSEUDOCODE
+entry: (ctx, e, { system, self }) => {
+  system.register('todo', self);
+}
+```
+
+**Getting listings**
+
+The `system` value is made available in the 3rd argument to action/guard/invoke creators, which contains the `system.receptionist` that extends the `ActorRef` interface.
+
+Listing all `ActorRef` objects registered under a key is done via `system.receptionist.list(key)`, which synchronously returns an array of `ActorRef` objects.
+
+```ts
+const todosMachine = createMachine({
+  // ...
+  on: {
+    'todos.logAll': {
+      actions: (ctx, e, { system }) => {
+        // Lists all todo ActorRefs
+        console.log(system.receptionist.list('todo'));
+      }
+    },
+    'todos.completeAll': {
+      actions: pure((ctx, e, { system }) => {
+        return system.receptionist.list('todo').map(todoRef => {
+          return sendTo(todoRef, { type: 'complete' });
+        });
```

Comment:
unregistering would be easier if we'd keep it internally as a Set and I don't think there is a strong reason to do an implicit conversion to an array on behalf of the caller

Reply:
How would you do the "send this event to all todo actors" use-case, then?